js-api-spec/compiler.node.test.ts

```diff
@@ -0,0 +1,142 @@
+
+  beforeEach(() => {
+    compiler = initCompiler();
+  });
+
+  afterEach(() => {
+    compiler.dispose();
+  });
+
+  describe('compile', () => {
+    it('performs complete compilations', () =>
+      sandbox(dir => {
+        const logger = getLogger();
+        dir.write({'input.scss': '@import "bar"; .fn {value: foo(bar)}'});
+        const result = compiler.compile(dir('input.scss'), {
+          importers,
+          functions,
+          logger,
+        });
+        expect(result.css).toEqualIgnoringWhitespace(
+          '.import {value: bar;} .fn {value: "bar";}'
+        );
+        expect(logger.debug).toHaveBeenCalledTimes(1);
+      }));
+
+    it('performs compilations in callbacks', () =>
+      sandbox(dir => {
+        dir.write({'input-nested.scss': 'x {y: z}'});
+        const nestedImporter = {
+          canonicalize: (url: string) => new URL(`u:${url}`),
+          load: (url: typeof URL) => ({
+            contents: compiler.compile(dir('input-nested.scss')).css,
+            syntax: 'scss' as const,
+          }),
+        };
+        dir.write({'input.scss': '@import "nested"; a {b: c}'});
+        const result = compiler.compile(dir('input.scss'), {
+          importers: [nestedImporter],
+        });
+        expect(result.css).toEqualIgnoringWhitespace('x {y: z;} a {b: c;}');
+      }));
+
+    it('throws after being disposed', () =>
+      sandbox(dir => {
+        dir.write({'input.scss': '$a: b; c {d: $a}'});
+        compiler.dispose();
+        expect(() => compiler.compile(dir('input.scss'))).toThrowError();
+      }));
+  });
+});
+
+describe('AsyncCompiler', () => {
+  let compiler: AsyncCompiler;
+  const runs = 1000; // Number of concurrent compilations to run
+
+  // A slow importer that executes a callback after a delay
+  const getSlowImporter = (callback: () => void) => ({
+    canonicalize: async (url: string) => new URL('foo:bar'),
+    load: async (url: typeof URL) => {
+      await new Promise(resolve => setTimeout(resolve, 100));
+      callback();
+      return {contents: '', syntax: 'scss' as const};
+    },
+  });
+
+  beforeEach(async () => {
+    compiler = await initAsyncCompiler();
+  });
+
+  afterEach(async () => {
+    await compiler.dispose();
+  });
+
+  describe('compileAsync', () => {
+    it('handles multiple concurrent compilations', () =>
+      sandbox(async dir => {
+        const logger = getLogger();
+        const compilations = Array(runs)
+          .fill(0)
+          .map((_, i) => {
+            const filename = `input-${i}.scss`;
+            dir.write({
+              [filename]: `@import "${i}"; .fn {value: foo(${i})}`,
+            });
+            return compiler.compileAsync(dir(filename), {
+              importers: asyncImporters,
+              functions,
+              logger,
+            });
+          });
+        Array.from(await Promise.all(compilations))
+          .map((result: CompileResult) => result.css)
+          .sort(sortCompiled)
```

Comment:
@jerivas What's the purpose of sorting here? I think that `Promise.all` will maintain the order, and sorting could hide if the first compilation is returning the results of the second compilation.

Reply:
Huh, I recall getting errors in the test at some point due to ordering issues, but you are correct; I just tried it and it's not needed. Thanks for catching that!

Reply:
Fixed in a2e6f0cd1f141522ad9a592953d9cd68931ad44f